ext/backtracie_native_extension/public/backtracie.h

```diff
@@ -0,0 +1,187 @@
+  // 1 -> ruby frame / 0 -> cfunc frame
+  uint32_t is_ruby_frame : 1;
+  // 1 means self really is the self object, 0 means it's the class of the self.
+  uint32_t self_is_real_self : 1;
+  // The iseq & cme; one, both, or neither might actually be available; if
+  // they're not, they will be Qnil. These need to be GC marked if you wish to
+  // keep the location alive.
+  VALUE iseq;
+  VALUE callable_method_entry;
+  // This is either self, or rb_class_of(self), depending on a few thing.
+  // It will be the actual self object for the frame if
+  //   - self is the toplevel binding,
+  //   - self is rb_mRubyVMFrozenCore,
+  //   - self is a class or module
+  // Otherwise, it will be rb_class_of(self).
+  // This is done so that, should the caller decide to retain the raw_location
+  // instance for a while, GC mark its VALUEs, and stringify the backtrace
+  // later, we both:
+  //   - retain enough information to actually produce a good name for the
+  //   method,
+  //   - but also don't hold onto random objects that would otherwise be GC'd
+  //   jsut
+  //     because they happened to be in a backtrace.
+  // This needs to be marked if you wish to keep the location alive
+  VALUE self_or_self_class;
+  // Raw PC pointer; not much use to callers, but saved so we can calculate the
+  // lineno later on.
+  const void *pc;
+} raw_location;
+
+// Returns the number of Ruby frames currently live on the calling thread; this
+// can be used to judge how many times backtracie_capture_frame_for_thread()
+// should be called to capture the actual frames.
+BACKTRACIE_API int backtracie_frame_count_for_thread(VALUE thread);
+// Equivalent to backtracie_frame_count_for_thread(rb_thread_current())
+BACKTRACIE_API int backtracie_frame_count();
+// Fills in the raw_location struct with the details of the Ruby call stack
+// frame on the given thread at the given index. The written frame may or may
+// not be actually meaningful; if it is not, then loc->is_valid will be 0.
+// loc->abi_version MUST be filled in with the value of BACKTRACIE_ABI_VERSION
+// before passing it in, to ensure ABI compatability. The frame index 0 is "the
+// most recently called code" - the bottom of the stack, if the stack is
+// considered to grow downwards. Returns true if there might be more frames
+// after this one, and false otherwise. The intended usage for this API is
+// something like this:
+//
+//   VALUE thread = rb_thread_current();
+//   int frame_count = backtracie_frame_count_for_thread(thread);
+//   raw_location *locs = malloc(sizeof(raw_location) * frame_count);
+//   bool have_more;
+//   int i = 0;
+//   do {
+//     locs[i].abi_version = BACKTRACIE_ABI_VERSION;
+//     have_more = backtracie_capture_frame_for_thread(thread, i, &locs[i]);
+//     if (locs[i].is_valid) { i++; }
+//  } while (have_more);
+BACKTRACIE_API
+bool backtracie_capture_frame_for_thread(VALUE thread, int frame_index,
+                                         raw_location *loc);
```

Comment:
embedding `abi_version` in the struct was inspired by APIs that require you to put sizeof(struct) in the struct. It could work as an argument too, but you'd have to pass it to _every_ function that accepted a raw_location, in theory.

We could also just bin it, and when the time comes when we want to promise ABI stability, we declare that as the "v1 ABI". Then, if we need to make an ABI break, we could use symbol versioning to do so? Or even just make a new family of functions `backtracie2_*` and keep them alongside the v1 functions.

tl;dr - the ABI versioning thing is probably complexity we don't need right now. I explored the idea because I wasn't sure what the implications of having Ruby extensions expose symbols to each other actually were, but I think I've convinced myself it can be solved in some way when the time comes to do so.

as for valid - my original rationale was that it's handy to have in the structure itself so you can iterate through an array and skip the invalid ones, but actually in hindsight it seems that since I settled on the caller constructing the array, and so the caller can just overwrite invalid frames so there are no gaps.

Reply:
Yeah I think I also prefer cleaning up the ABI stuff until we put out v1; and then we can always bump for v2 as needed. (It's one of those good problems to have -- enough people using backtracie for that to be a factor! :D )